src/pcp.rs

```diff
@@ -0,0 +1,539 @@
+/// determines whether a given value is valid.
+pub trait Value<F, G>: Sized
+where
+    F: FieldElement,
+    G: Gadget<F>,
+{
+    /// Evalauates the validity circuit on the given input (i.e., `self`) and returns the output.
+    /// Slice `rand` is the random input consumed by the validity circuit.
+    ///
+    /// ```
+    /// use prio::pcp::types::Boolean;
+    /// use prio::pcp::Value;
+    /// use prio::field::{FieldElement, Field64};
+    ///
+    /// type F = Field64;
+    /// type T = Boolean<F>;
+    ///
+    /// let x = T::new(false);
+    ///
+    /// let mut rand: Vec<F> = Vec::with_capacity(x.valid_rand_len());
+    /// for _ in 0..rand.len() {
+    ///     rand.push(F::rand());
+    /// }
+    ///
+    /// let v = x.valid(&mut x.gadget(0), &rand).unwrap();
+    /// assert_eq!(v, F::zero());
+    /// ```
+    fn valid(&self, g: &mut dyn GadgetCallOnly<F>, rand: &[F]) -> Result<F, PcpError>;
+
+    /// Returns a reference to the underlying data.
+    fn as_slice(&self) -> &[F];
+
+    /// Constructs an instance of this type from the given data without checking whether the data
+    /// is valid. This method takes ownership of `data`.
+    fn clone_with(&self, data: Vec<F>) -> Self;
+
+    /// The length of the random input expected by the validity circuit.
+    fn valid_rand_len(&self) -> usize;
+
+    /// The number of calls to the gadget made when evaluating the validity circuit.
+    fn valid_gadget_calls(&self) -> usize;
+
+    /// Returns an instance of the gadget associated with the validity circuit. `in_len` is the
+    /// maximum degree of the polynomials passed into `call_poly`. If `call_poly` is never used by
+    /// then it is safe to set `in_len == 0`.
+    fn gadget(&self, in_len: usize) -> G;
+}
+
+/// The gadget functionality required for evaluating a validity circuit. The `Gadget` trait
+/// inherits this trait.
+pub trait GadgetCallOnly<F: FieldElement> {
+    /// Evaluates the gadget on input `inp` and returns the output.
+    fn call(&mut self, inp: &[F]) -> Result<F, PcpError>;
+
+    /// Returns the circuit's arity, i.e., the expected length of the input to `call`.
+    fn call_in_len(&self) -> usize;
+}
+
+/// The sub-circuit associated with some validity circuit. A gadget is called either on a sequence
+/// of finite field elements or a sequence of polynomials over a finite field.
+pub trait Gadget<F: FieldElement>: GadgetCallOnly<F> {
+    /// Evaluate the gadget on input of a sequence of polynomials. The output is written to `outp`.
+    fn call_poly<V: AsRef<[F]>>(&mut self, outp: &mut [F], inp: &[V]) -> Result<(), PcpError>;
+
+    /// The maximum degree of the polynomial output by `call_poly` as a function of the maximum
+    /// degree of each input polynomial.
+    fn call_poly_out_len(&self, in_len: usize) -> usize;
+}
+
+/// Generate a proof of an input's validity.
+pub fn prove<F, G, T>(x: &T) -> Result<Proof<F>, PcpError>
+where
+    F: FieldElement,
+    G: Gadget<F>,
+    T: Value<F, G>,
```

Comment:
nit: Now that it's `Value` instead of `Type`, it'd be nice for this (and lots of other generic declarations) to do `V: Value<F, G>`.

Reply:
Done.